### PR TITLE
Adds the Dark Matt-eor when you emag a stupid amount of meteor shields + lots of meteor file sorting + qol + dark matter singularity + dark matt-eor summoning final traitor objective

### PR DESCRIPTION
**Pull request by tralezab**
## About The Pull Request

<details>
  <summary>Dark Matt-eor Image</summary>
  
  image

</details>

> A barely visible blur in the cosmic darkness, like a ghostly shadow on a moonless night. A piercing howl in the vacuum of space, as if it were tearing the fabric of reality. A twisted halo of light around it, bending and breaking the rays of distant suns. A shower of quantum sparks, flickering and fading in its wake. A dark matter meteor (dark matt-eor) is a wonder to witness, and to dread.

> A sudden impact, like a hammer blow to the heart of the station. A violent tremor, shaking and shattering the metal walls and windows. A deafening roar, as the air rushes out of the breached hull. A blinding flash, as the dark matter meteor unleashes its hidden energy. A tiny black hole, forming and growing in the center of the station. A relentless pull, dragging everything towards the abyss. A dark matter meteor is incredibly deadly.

Emagging too many meteor shields will summon a dark matt-eor. This comes with several warnings, and after awhile, warns the station that someone is trying to summon a dark matteor.

The dark matt-eor itself is not that damaging in its impact, but drops a singularity in its final resting place.

## Why It's Good For The Game

It's a new way to terrorize a round as an antagonist. Before, emagging a lot of meteor shields would basically make meteor showers the only event that can run, which is cool, but since constant meteor waves are going to destroy the station, let's also throw in the mother of all meteors!

This also adds warnings to spamming emagging meteor shields, which imo needs it. The round ends when someone spams emagged meteor shields, and since they're meteor shields nobody is going to reasonably check on them.

## Changelog
:cl:
add: The dark matt-eor
add: Summon a dark matt-eor final traitor objective
add: Dark matter singularity variant, which can't grow as big as a regular singularity but hungers for blood
code: cleaned up/sorted meteor shield code, satellite control, and more
qol: added a lot of feedback to interacting with meteor shields
balance: emagging a lot of meteor shields warns the station, but emagging enough of them summons a Dark Matt-eor.
/:cl:

### CODE REVIEW

File: code/modules/station_goals/meteor_shield.dm [hunk 1]
```diff
@@ -172,27 +79,61 @@
...
+
+/obj/machinery/satellite/meteor_shield/proc/handle_new_emagged_shield_threshold()
+	switch(highest_emagged_threshold_reached)
+		if(EMAGGED_METEOR_SHIELD_THRESHOLD_ONE)
+			say("Warning. Meteor strike probability entering dangerous ranges for more exotic meteors.")
+		if(EMAGGED_METEOR_SHIELD_THRESHOLD_TWO)
+			say("Warning. Risk of dark matter congealment entering existent ranges. Further tampering will be reported.")
+		if(EMAGGED_METEOR_SHIELD_THRESHOLD_THREE)
+			say("Warning. Further tampering has been reported.")
+			priority_announce()
```
Comment (Fikou) [inline code/modules/station_goals/meteor_shield.dm:121]: nice announcement

Comment (tralezab) [reply code/modules/station_goals/meteor_shield.dm:121]: oops i already fixed this locally...

File: code/game/machinery/satellite/satellite_control.dm [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	. = ..()
+	if(.)
+		return
+
+	switch(action)
+		if("toggle")
+			toggle(text2num(params["id"]))
+			. = TRUE
+
+/obj/machinery/computer/sat_control/proc/toggle(id)
```
Comment (Fikou) [inline code/game/machinery/satellite/satellite_control.dm:24]: mind touching this and the one below to use descriptive var names?

**Comment by flowercuco** [comment]
no one throws spaghetti at the wall like you tralezab thank you so much

**Comment by Fikou** [comment]
> When he designs a traitor final objective, but just puts it behind the emag instead...

just build a bsa to stop the singulo

**Comment by tralezab** [comment]
## I'm just enjoying watching the dang thing

image

**Comment by Fikou** [comment]
> Fikou when you can't order the BSA and Meteor Sats on the same round

yeah you can when its extended

**Comment by ATH1909** [comment]
what's the purpose of the warnings when you can just construct 10 meteor sats in a remote location and emag them all in a row to instantly summon a singularity meteor?

**Comment by ATH1909** [comment]
> ATHATH i _just_ said I was considering that. BRO

okay hear me out here

instead of making this an effect for emagging too many meteor shields, you make this an actual traitor final objective that can only be given if the station has a meteor shields objective. make it, like, a device you strap onto one of the shields or something.

**Comment by ATH1909** [comment]
> ATHATH i _just_ said I was considering that. BRO
> 
> image

o

**Comment by tralezab** [comment]
This is a neat final traitor objective, sure I'll add it as one since it's pretty adaptable into one

**Comment by Kubisopplay** [comment]
Please let me throw that as an admin smite

**Comment by tralezab** [comment]
Changes to avoid abuse and keep it a FUN feature:
- 1 minute cooldown on emagging
- always announces the dark matteor avoiding some edge cases where it collides before the announce can find it
- all hacked meteor shields give off a gps signal

**Comment by TheSmallBlue** [comment]
Should have like, a 10% chance of being named "The Big Chip" instead

**Comment by LemonInTheDark** [comment]
Ran out from sidewalk into 6 lane road/10
Pogged up

**Comment by DevinShafer** [comment]
> > When he designs a traitor final objective, but just puts it behind the emag instead...
> 
> just build a bsa to stop the singulo

Just, throw a bluespace bag at it, If the codes still there I believe it "shrinks" it still.
or you know, throw the clown, the clown has fun interaction

either way I am to blame for this PR to be existent and I love it

**Comment by Pumpkinoe** [comment]
copyrighted pumpkin0 2023 its my idea i swear

**Comment by Singul0** [comment]
> > ATHATH i _just_ said I was considering that. BRO
> 
> okay hear me out here
> 
> instead of making this an effect for emagging too many meteor shields, you make this an actual traitor final objective that can only be given if the station has a meteor shields objective. make it, like, a device you strap onto one of the shields or something.
> 
> this will satisfy the slow cook-enjoyers who just wanna bring back the meteors gamemode, still add the cool dark matter meteor to the game, and prevent people from raging over some doofus summoning a singulo 20 minutes into the round.

that assumes people actually do station goals. this will make the objective more annoying and harder to do than any other final objjectives since the traitor need to buy and build a meteorshield beforehand.

**Comment by nianjiilical** [comment]
who is dark matt and what have we done to incur his wrath

File: code/datums/components/singularity.dm [hunk 3]
```diff
@@ -335,5 +341,47 @@
...
+			continue
+		var/distance_from_target = get_dist(target, atom_parent)
+		if(distance_from_target < closest_distance)
+			closest_distance = distance_from_target
+			closest_target = target
+	return closest_target
+
+/// gives a little fluff warning that someone is being hunted.
+/datum/component/singularity/bloodthirsty/proc/foreboding_nosebleed(mob/living/target)
+	if(iscarbon(target))
```
Comment (Watermelon914) [inline code/datums/components/singularity.dm:372]: you should early return here

**Review by Watermelon914 — APPROVED** [review]

File: code/datums/components/singularity.dm [hunk 4]
```diff
@@ -335,5 +343,60 @@
...
+		return
+	var/mob/living/carbon/carbon_target = target
+	var/obj/item/bodypart/head = carbon_target.get_bodypart(BODY_ZONE_HEAD)
+	var/has_no_blood = HAS_TRAIT(carbon_target, TRAIT_NOBLOOD)
+	if(head)
+		if(has_no_blood)
+			to_chat(carbon_target, span_notice("You get a headache."))
+			return
+		head.adjustBleedStacks(5)
+		carbon_target.visible_message(span_notice("[carbon_target] gets a nosebleed."), span_notice("You get a nosebleed."))
```
Comment (Fikou) [inline code/datums/components/singularity.dm:395]:
```suggestion
		carbon_target.visible_message(span_notice("[carbon_target] gets a nosebleed."), span_warning("You get a nosebleed."))
```

File: code/game/machinery/satellite/satellite_control.dm [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+			toggle(text2num(params["id"]))
+			. = TRUE
+
+/obj/machinery/computer/sat_control/proc/toggle(toggled_id)
+	var/turf/current_turf = get_turf(src)
+	for(var/obj/machinery/satellite/satellite in GLOB.machines)
+		if(satellite.id == toggled_id && is_valid_z_level(get_turf(satellite), current_turf))
+			satellite.toggle()
+
+/obj/machinery/computer/sat_control/ui_data()
```
Comment (Fikou) [inline code/game/machinery/satellite/satellite_control.dm:30]: on this proc use descriptive var names too

File: code/game/objects/items/toys.dm [hunk 6]
```diff
@@ -295,6 +295,19 @@
...
+ */
+/obj/item/toy/spinningtoy/dark_matter
+	name = "dark matter singularity"
+	desc = "<i>\"Surviving the encounter with the \
+		horrible thing, I realized immediately what I \
+		had to do: sell marketable toys of it. \
+		\"</i><br>- Chief Engineer Ship Chafer"
+	icon = 'icons/obj/engine/singularity.dmi'
+	icon_state = "dark_matter_s1"
+	item_flags = NO_PIXEL_RANDOM_DROP
```
Comment (Fikou) [inline code/game/objects/items/toys.dm:309]: inherited from parent

File: code/modules/meteors/meteor_dark_matteor.dm [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+/obj/effect/meteor/dark_matteor/shield_defense(obj/machinery/satellite/meteor_shield/defender)
+	defender.visible_message(span_danger("[defender]'s beam is reflected by [src]!"))
+	new /obj/effect/temp_visual/explosion/fast(get_turf(defender))
+	qdel(defender)
+	return FALSE
+
+/obj/effect/meteor/dark_matteor/handle_stopping()
+	. = ..()
+	SSsecurity_level.set_level(previous_security_level)
```
Comment (Fikou) [inline code/modules/meteors/meteor_dark_matteor.dm:63]: check if its still on red here

File: code/modules/meteors/meteor_dark_matteor.dm [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	var/atom/movable/warp_effect/warp
+	/// and another oh shit in the form of quantum sparks
+	var/datum/effect_system/spark_spread/quantum/spark_system
+	/// in case we miss, we can go back to the previous security level
+	var/previous_security_level
+
+/obj/effect/meteor/dark_matteor/Initialize(mapload, turf/target)
+	. = ..()
+	previous_security_level = SSsecurity_level.get_current_level_as_number()
+	SSsecurity_level.set_level(SEC_LEVEL_RED)
```
Comment (Fikou) [inline code/modules/meteors/meteor_dark_matteor.dm:24]: what happens if its already red

**Review by Fikou — CHANGES_REQUESTED** [review]

**Comment by Loafin34** [comment]
this is the first maia roentgen reference in code. very surprising.

**Comment by carshalash** [comment]
There's seriously no warning or anything for this?

**Comment by Draggeru** [comment]
> There's seriously no warning or anything for this?

There is unless an admin forces it.

**Pull request merged**